python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,471 @@
+                # test yet) and we are not ignoring prereqs, remove it.
+                #
+                # Note 1: This test may depend on other tests that were also skipped, so its possible to hit
+                #         this method multiple times. Hence the check for an initialized status
+                #
+                # Note 2: Dry Run is here to override the tester setting this to PASS, because we still want
+                #         to know when tests like this would normally be skipped due to a dependency issue
+                #         (instead of printing out a green passing 'DRY RUN' status)
+                if (dwn_tester.isInitialized() and not dwn_tester.skipPrereqs(self.options)) \
+                   or self.options.dry_run:
+                    # Set the status and assign the skipped bucket
+                    dwn_tester.setStatus('skipped dependency', dwn_tester.bucket_skip)
+
+                    # Add this deleted tester to a set
+                    deleted_tester_set.add(dwn_tester)
+
+                    # Ask a status thread to print our results
+                    self.statusGo(downstream_test_name)
+
+                # Like Note 1 above, the same applies here. We need to _attempt_ to delete the node, but not
+                # tell the harness multiple times that this test was skipped. This 'downsteam' node just
+                # happens to exist in mulitple places in the graph. Exp:
+                # A     B
+                #  \   /
+                #    C  <--- Downstream 'C' node mentioned in both A and B.
+                if not dwn_tester.skipPrereqs(self.options):
+                    dag_object.delete_node_if_exists(downstream_test_name)
+
+        # Return the set of tester objects we deleted
+        return deleted_tester_set
+
+    # Loop through the testers and add them to the runner pool for execution
+    def schedule(self, testers):
+        # We didn't recieve anything... pesky TestHarness
+        if testers == []:
+            return
+
+        # Immediately test for race conditions and return if condition exists
+        if self.checkRaceCondition(testers):
+            return
+
+        # Local storage of skipped tester objects
+        skipped_or_failed_tests = set([])
+
+        # Local storage of downstream deleted tester objects
+        downstream_deleted = set([])
+
+        # Local DAG
+        tester_dag = dag.DAG()
+
+        # Create the DAG, skipped_testers, and just about anything we can during this iteration loop
+        for tester in testers:
+            with self.thread_lock:
+                # Create tester name to object map
+                self.scheduled_name_to_object[tester.getTestName()] = tester
+
+            # Create a node for this tester if not already exists (prereq iteration creates nodes)
+            if not tester_dag.node_exists(tester.getTestName()):
```

Comment:
Another idea for you. You might add another method to the DAG that just adds nodes all the time so you don't have to check first. If you find yourself always doing this, then it would be useful to have that utility.